### PR TITLE
[MRG] Linear Neural Networks/Linear Regression Implementation from Scratch

### PR DESCRIPTION
**Pull request by xiaotinghe**

### CODE REVIEW

**Comment by mli** [comment]
Job d2l-zh/PR-553/1 is complete. 
Check the results at http://preview.d2l.ai/d2l-zh/PR-553/

**Comment by mli** [comment]
Job d2l-zh/PR-553/2 is complete. 
Check the results at http://preview.d2l.ai/d2l-zh/PR-553/

**Comment by mli** [comment]
Job d2l-zh/PR-553/3 is complete. 
Check the results at http://preview.d2l.ai/d2l-zh/PR-553/

File: chapter_linear-networks/linear-regression-scratch.md [hunk 1]
```diff
@@ -1,7 +1,12 @@
 # 线性回归的从零开始实现
 :label:`sec_linear_scratch`
 
-现在您已经了解了线性回归背后的关键想法，我们可以开始在代码中实现实践。在本节中，我们将从头开始实现整个方法，包括数据流水线、模型、损失函数和微型批次随机梯度下降优化器。虽然现代深度学习框架几乎可以自动执行所有这些工作，但从头开始实施操作是确保您真正了解自己在做什么的唯一方法。此外，当涉及到自定义模型时，定义我们自己的层或损失功能时，了解事情如何在引擎盖下工作将被证明是方便的。在本节中，我们将仅依靠张量和自动分化。之后，我们将利用深度学习框架的钟声和口哨，介绍一个更简洁的实现。
+在了解线性回归的关键思想后，我们可以开始在代码中动手实现了。
+
```
Comment (astonzhang) [inline chapter_linear-networks/linear-regression-scratch.md:5]: 这里不要添加新空行，会让下面的文字另起一段

File: chapter_linear-networks/linear-regression-scratch.md [hunk 2]
```diff
@@ -29,13 +34,18 @@ import random
 
 ## 生成数据集
 
-为了保持简单，我们将根据带有加法噪声的线性模型构建一个人工数据集。我们的任务是使用我们的数据集中包含的有限示例来恢复此模型的参数。我们将保持低维数据，以便我们可以轻松地显示数据。在以下代码片段中，我们生成一个包含 1000 个示例的数据集，每个示例由从标准正态分布采样的 2 个要素组成。因此，我们的综合数据集将是一个矩阵 $\mathbf{X}\in \mathbb{R}^{1000 \times 2}$。
+为了简单起见，我们将根据带有噪声的线性模型构造一个人工数据集。
+我们的任务是使用有限样本的数据集来恢复这个模型的参数。
+我们将使用低维数据，这样我们可以很容易地将其可视化。
+
```
Comment (astonzhang) [inline chapter_linear-networks/linear-regression-scratch.md:40]: 这里不要空行。

File: chapter_linear-networks/linear-regression-scratch.md [hunk 3]
```diff
@@ -29,13 +34,18 @@ import random
...
+在下面的代码中，我们生成一个包含1000个样本的数据集，每个样本包含从标准正态分布中采样的2个特征。我们的合成数据集是一个矩阵 $\mathbf{X}\in \mathbb{R}^{1000 \times 2}$。
 
-生成我们的数据集的真实参数将是 $\mathbf{w} = [2, -3.4]^\top$ 和 $b = 4.2$，我们的合成标注将根据以下线性模型进行分配，其噪声项为 $\epsilon$：
+我们使用线性模型参数$\mathbf{w} = [2, -3.4]^\top$、$b = 4.2$和噪声项$\epsilon$生成数据集及其标签:
```
Comment (astonzhang) [inline chapter_linear-networks/linear-regression-scratch.md:43]: 顿号改回“和”。一般并列标点规范是：
“甲和乙”
“甲、乙和丙”
“甲、乙、丙和丁”

File: chapter_linear-networks/linear-regression-scratch.md [hunk 4]
```diff
@@ -113,9 +125,10 @@ def data_iter(batch_size, features, labels):
         yield tf.gather(features, j), tf.gather(labels, j)
 ```
 
-一般而言，请注意，我们希望使用大小合理的小批量来利用 GPU 硬件，这在并行化操作方面非常出色。由于每个样本都可以并行通过我们的模型提供，并且每个样本的损失函数的梯度也可以并行进行，因此 GPU 允许我们处理数百个样本的时间比处理单个示例所花费的时间少得多。
+一般来说，我们使用合理大小的小批量来利用GPU硬件的优势，因为GPU在并行处理方面表现出色。每个样本都可以并行地通过我们的模型计算，并且每个样本的损失函数的梯度也可以被并行地计算，GPU可以让我们处理几百个样本不比处理一个样本所花费的时间多太多。
```
Comment (astonzhang) [inline chapter_linear-networks/linear-regression-scratch.md:128]: “GPU可以让我们处理几百个样本不比处理一个样本所花费的时间多太多”

这句话比较拗口

File: chapter_linear-networks/linear-regression-scratch.md [hunk 5]
```diff
@@ -113,9 +125,10 @@ def data_iter(batch_size, features, labels):
...
 
-为了建立一些直觉，让我们阅读并打印第一批量数据示例。每个微型批次中要素的形状告诉我们微型批次大小和输入要素的数量。同样，我们的小批标签将采用 `batch_size` 给出的形状。
+我们读取第一个小批量数据样本并打印。每个批量的特征维度说明了批量大小和输入特征数。
```
Comment (astonzhang) [inline chapter_linear-networks/linear-regression-scratch.md:130]: 这个地方是不是省略掉太多原文了？

File: chapter_linear-networks/linear-regression-scratch.md [hunk 6]
```diff
@@ -152,13 +169,15 @@ w = tf.Variable(tf.random.normal(shape=(2, 1), mean=0, stddev=0.01),
 b = tf.Variable(tf.zeros(1), trainable=True)
 ```
 
-初始化我们的参数后，我们的下一个任务是更新它们，直到它们充分适合我们的数据。每次更新都需要采用我们的损失函数相对于参数的梯度。鉴于这个梯度，我们可以更新每个参数，以减少损失的方向。
-
-由于没有人愿意明确计算梯度（这很乏味，容易出误差），我们使用 :numref:`sec_autograd` 中引入的自动差异来计算梯度。
+在初始化参数之后，我们的下一个任务是更新这些参数，直到这些参数足够拟合我们的数据。
+每次更新都需要计算损失函数有关参数的梯度。有了这个梯度，我们就可以向减小损失的方向更新每个参数。
+因为手动计算梯度很乏味而且容易出错，所以没有人会手动计算梯度。我们使用 :numref:`sec_autograd` 中引入的自动微分来计算梯度。
```
Comment (astonzhang) [inline chapter_linear-networks/linear-regression-scratch.md:174]: 乏味 -》 枯燥

File: chapter_linear-networks/linear-regression-scratch.md [hunk 7]
```diff
@@ -180,9 +201,12 @@ def squared_loss(y_hat, y):  #@save
 
 ## 定义优化算法
 
-正如我们在 :numref:`sec_linear_regression` 中讨论的那样，线性回归有一个封闭式解。然而，这不是一本关于线性回归的书：它是一本关于深度学习的书。由于本书介绍的其他模型都无法通过分析解决，因此我们将借此机会介绍您第一个微型批次随机梯度下降的工作样本。
+正如我们在 :numref:`sec_linear_regression` 中讨论的，线性回归有解析解。然而，这是一本关于深入学习的书，而不是一本关于线性回归的书。
+由于这本书介绍的其他模型都没有解析解，我们将利用这个机会介绍第一个小批量随机梯度下降的工作示例。
 
-在每一步，使用从我们的数据集中随机绘制的一个微型批次，我们将估计相对于我们的参数的损失梯度。接下来，我们将更新我们的参数，以减少损失的方向。以下代码应用微批次随机梯度下降更新，给定一组参数、学习率和批量大小。更新步骤的大小由学习率 `lr` 决定。因为我们的损失是以小批次示例的总和计算的，所以我们按批批量大小 (`batch_size`) 标准化我们的步长大小，这样一个典型步长大小的幅度不会在很大程度上取决于我们对批次大小的选择。
+在每一步中，使用从数据集中随机抽取的一个小批量，然后根据参数计算损失的梯度。接下来，朝着减少损失的方向更新我们的参数。
+下面的代码使用了小批量随机梯度下降更新，接受模型参数集合、学习速率和批量大小作为输入。每一步更新的大小由学习速率`lr`决定。
```
Comment (astonzhang) [inline chapter_linear-networks/linear-regression-scratch.md:208]: “（主语）接受模型参数集合、学习速率和批量大小作为输入。”
缺主语

File: chapter_linear-networks/linear-regression-scratch.md [hunk 8]
```diff
@@ -208,20 +232,22 @@ def sgd(params, grads, lr, batch_size):  #@save
...
 ```
 
-## 培训
-
-现在我们已经完成了所有部件，我们已经准备好实施主要的训练循环。理解此代码至关重要，因为在深度学习的整个职业生涯中，您将一遍又一遍地看到几乎相同的训练循环。
+## 训练
 
-在每次迭代中，我们将获取一小批训练示例，并通过我们的模型传递它们以获得一组预测。计算损失后，我们启动向后通过网络，存储相对于每个参数的渐变。最后，我们将调用优化算法 `sgd` 来更新模型参数。
+现在我们已经准备好了所有的部分，可以实现主要的训练循环部分了。
+理解这段代码至关重要，因为在深度学习的整个职业生涯中，你会看到一遍又一遍几乎相同的训练循环。
```
Comment (astonzhang) [inline chapter_linear-networks/linear-regression-scratch.md:238]: 在整个深度学习的职业生涯中

File: chapter_linear-networks/linear-regression-scratch.md [hunk 9]
```diff
@@ -208,20 +232,22 @@ def sgd(params, grads, lr, batch_size):  #@save
...
+在每次迭代中，我们读取一小批量训练样本，并通过我们的模型来获得一组预测。
+计算完损失后，我们开始反向通过网络，存储每个参数的梯度。最后，我们调用优化算法 `sgd` 来更新模型参数。
```
Comment (astonzhang) [inline chapter_linear-networks/linear-regression-scratch.md:240]: 反向通过网络 -》 反向传播（这是术语）

File: chapter_linear-networks/linear-regression-scratch.md [hunk 10]
```diff
@@ -208,20 +232,22 @@ def sgd(params, grads, lr, batch_size):  #@save
...
 * 初始化参数
-* 重复直到完成
-    * 计算梯度
-    * 更新参数
+* 重复，直到完成
+    * 计算梯度 $\mathbf{g} \leftarrow \partial_{(\mathbf{w},b)} \frac{1}{|\mathcal{B}|} \sum_{i \in \mathcal{B}} l(\mathbf{x}^{(i)}, y^{(i)}, \mathbf{w}, b)$
+    * 更新参数 $(\mathbf{w}, b) \leftarrow (\mathbf{w}, b) - \eta \mathbf{g}$
 
-在每个 *epoch* 中，我们将遍历整个数据集（使用 `data_iter` 函数），一旦遍历训练数据集中的每个样本（假设示例数量可以被批量次大小除去）。时代的数量和学习率 `lr` 都是超参数，我们在这里分别设置为 3 和 0.03。不幸的是，设置超参数很棘手，需要通过试验和误差进行一些调整。我们现在保留这些细节，但稍后在 :numref:`chap_optimization` 中进行修改。
+在每个*迭代周期（epoch）*中，我们使用 `data_iter` 函数遍历整个数据集，并将训练数据集中所有样本都使用一次（假设样本数能够被批量大小整除）。这里的迭代周期个数num_epochs和学习率lr都是超参数，分别设3和0.03。设置超参数很棘手，需要通过反复试验进行调整。
```
Comment (astonzhang) [inline chapter_linear-networks/linear-regression-scratch.md:249]: num_epochs和lr缺了``

File: chapter_linear-networks/linear-regression-scratch.md [hunk 11]
```diff
@@ -272,30 +298,33 @@ for epoch in range(num_epochs):
...
 ## 练习
 
 1. 如果我们将权重初始化为零，会发生什么。算法仍然有效吗？
-1. 假设你是 [Georg Simon Ohm](https://en.wikipedia.org/wiki/Georg_Ohm) 试图拿出电压和电流之间的模型。您可以使用自动分化来学习模型的参数吗？
-1. 您可以使用 [普朗克定律](https://en.wikipedia.org/wiki/Planck%27s_law) 使用光谱能量密度来确定物体的温度吗？
-1. 如果你想计算二阶导数，你可能会遇到什么问题？你会如何解决这些问题？
-1.  为什么在 `squared_loss` 函数中需要使用 `reshape` 函数？
-1. 使用不同的学习速率进行实验，以了解损失函数值下降的速度。
-1. 如果示例数量不能除以批量大小，则 `data_iter` 函数的行为会发生什么情况？
+1. 假设你是 [Georg Simon Ohm](https://en.wikipedia.org/wiki/Georg_Ohm) 试图在电压和电流之建立一个模型。你能使用自动微分来学习模型的参数吗?
```
Comment (astonzhang) [inline chapter_linear-networks/linear-regression-scratch.md:322]: "假设你是 [Georg Simon Ohm]"后面加逗号

Comment (astonzhang) [reply chapter_linear-networks/linear-regression-scratch.md:322]: 本书需要对人名直接翻译

**Comment by mli** [comment]
Job d2l-zh/PR-553/4 is complete. 
Check the results at http://preview.d2l.ai/d2l-zh/PR-553/

**Comment by mli** [comment]
Job d2l-zh/PR-553/5 is complete. 
Check the results at http://preview.d2l.ai/d2l-zh/PR-553/

**Comment by mli** [comment]
Job d2l-zh/PR-553/6 is complete. 
Check the results at http://preview.d2l.ai/d2l-zh/PR-553/

**Comment by mli** [comment]
Job d2l-zh/PR-553/7 is complete. 
Check the results at http://preview.d2l.ai/d2l-zh/PR-553/
